src/common/TrackedOp.h

```diff
@@ -386,4 +279,156 @@ class TrackedOp : public boost::intrusive::list_base_hook<> {
+   * @param check a function consuming tracked ops, the function returns
+   *              false if it don't want to be fed with more ops
+   * @return True if there are any Ops to warn on, false otherwise
+   */
+  bool visit_ops_in_flight(utime_t* oldest_secs,
+			   std::function<bool(TrackedOp&)>&& visit);
+  /**
+   * walk through slow ops in flight
+   *
+   * @param[out] oldest_sec the amount of time since the oldest op was initiated
+   * @param[out] num_slow_ops total number of slow ops
+   * @param on_warn a function consuming tracked ops, the function returns
+   *                false if it don't want to be fed with more ops
+   * @return True if there are any Ops to warn on, false otherwise
+   */
+  bool with_slow_ops_in_flight(utime_t* oldest_secs,
+			       int* num_slow_ops,
+			       std::function<void(TrackedOp&)>&& on_warn);
+  /**
+   * Look for Ops which are too old, and insert warning
+   * strings for each Op that is too old.
+   *
+   * @param summary[out] a string summarizing slow Ops.
+   * @param warning_strings[out] A vector<string> reference which is filled
+   * with a warning string for each old Op.
+   * @param slow[out] total number of slow ops
+   * @return True if there are any Ops to warn on, false otherwise.
+   */
+  bool check_ops_in_flight(std::string* summary,
+			   std::vector<string> &warning_strings,
+			   int* slow = nullptr);
+
+  void on_shutdown() {
+    history.on_shutdown();
+  }
+  ~OpTracker();
+
+  template <typename T, typename U>
+  typename T::Ref create_request(U params)
+  {
+    typename T::Ref retval(new T(params, this));
+    retval->tracking_start();
+
+    if (is_tracking()) {
+      retval->mark_event("header_read", params->get_recv_stamp());
+      retval->mark_event("throttled", params->get_throttle_stamp());
+      retval->mark_event("all_read", params->get_recv_complete_stamp());
+      retval->mark_event("dispatched", params->get_dispatch_stamp());
+    }
+
+    return retval;
+  }
+};
+
+/* Defining the methods outside the TrackedOp to avoid circular
+ * dependency with OpTracker while preserving their inlineability.
+ * The `inline` marking allows to bypass the ODR rule:
+ *
+ *   "3.2/3 Every program shall contain exactly one definition
+ *   of every non-inline function or object that is used in that
+ *   program"
+ *
+ * See: https://stackoverflow.com/questions/7833941/putting- \
+        function-definitions-in-header-files#comment9553758_7834555
+ */
+inline void TrackedOp::tracking_start() {
+  if (tracker->register_inflight_op(this)) {
+    events.emplace_back(initiated_at, "initiated");
+    state = STATE_LIVE;
+  }
+}
+
+inline void TrackedOp::put() {
+  if (nref.load() == 1) {
+    switch (state.load()) {
+    case STATE_UNTRACKED:
+      --nref;
+      _unregistered();
+      delete this;
+      break;
+
+    case STATE_LIVE:
+      mark_event("done");
+      tracker->unregister_inflight_op(
+	TrackedOpRef(this, false /* = add_ref */));
```

Comment:
So we've added several branch cases to avoid decrement-and-increment when we unregister an op?
Guess the testing shows it's faster, just...kinda unpleasant.

Reply:
Yes, we have a trade-off here: more chances for control flow mis-speculation in exchange for avoiding atomic modifications. I think the second is greater evil as it hurts scalability while the cost is variable and can be really significant. Following factors contribute:

* The pure `lock` cost. Atomic RMW is costlier even in  a single threaded environment. These instructions have long latency (~20 cycles; comparable to a pipeline-refill), require microsequencer, translates into huge number of micro-ops etc.
* Cache line ping pong. This can affect even non-atomic members placed on the 64 bytes-aligned block (false sharing).
* On x86 a `locked` instruction imposes full memory barrier. Practically, this translates into draining all store and load buffers in the core before it can start any further memory operation.

It might be worth to put a comment linking [the very good article about atomics and contention](https://fgiesen.wordpress.com/2014/08/18/atomics-and-contention/). Personally, I found it's very useful.

Reply:
I'm sure you're right, some of the changesets just take a while to figure out that's why we're making the change, and then accepting that bus locks/memory flushes *yes really it is that expensive you wouldn't not believe just do it*. :D